### DISCUSSION TITLE
How to allow external users to access certain tables in my DB?

### DISCUSSION THREAD

**Discussion by gvko** [supabase/supabase#28012] [Questions]
I have two tables in my Supabase DB, called `users` and `tweets`. I've created a new role called `external_user_twitter_data` with "user can sign in" permission.

Then, I'm trying to create a policy for my tables. I need all kind of access to the tables for the given users: read, write, update, etc.

So, this is the query I'm trying to create the policy with:
```sql
create policy "Read-Write Twitter users for external users"
on "public"."users"
as PERMISSIVE
for ALL
to external_user_twitter_data
using (
(select auth.uid()) = user_id
)
with check (
(select auth.uid()) = user_id
);
```

as described in the [documentation here](https://supabase.com/docs/guides/database/postgres/row-level-security#update-policies).

But I'm getting an error: `ERROR:  column "user_id" does not exist`, which kind of makes sense, because I don't have `user_id` column on these tables. So, instead, should I create a Supabase user and put the user ID instead of `user_id` ? F.e. `select auth.uid()) IN ('some-user-id')` ?
Also, then, how am I supposed to assign the role `external_user_twitter_data` to a given user?

See the screenshot below of how I'm trying to do it in the Supabase admin dashboard.
<img width="644" alt="Screenshot 2024-07-15 at 19 24 11" src="https://github.com/user-attachments/assets/3bb57a1d-8b6f-4e1e-8f02-08566925ac9b">

**Comment by GaryAustin1** [comment]
Normally Supabase suggests for roles to use a role table and/or the jwt with a roles claim with auth-hooks https://supabase.com/docs/guides/database/postgres/custom-claims-and-role-based-access-control-rbac or app metadata in the jwt https://github.com/supabase-community/supabase-custom-claims .

What you are trying to do is more complex by using a Postgres role.  For that method would need to make the user obey RLS and use the TO part of the policy to limit.  Or you could use Postgres grants on the user.  The way to make a user that role is to change the role column in auth.users.  Supabase discourages this approach.  You also need to then grant the role to authenticator (the REST API role).
